torch/csrc/dynamo/eval_frame.c

```diff
@@ -786,18 +1082,42 @@ static PyObject* set_eval_frame_py(PyObject* dummy, PyObject* callback) {
   return set_eval_frame(callback, PyThreadState_GET());
 }
 
+static void destroy_cache_entry_wrapper(PyObject* code) {
+  PyObject* extra = NULL;
+  _PyCode_GetExtra((PyObject*)code, cache_entry_extra_index, (void*)&extra);
+  if (extra != NULL && extra != SKIP_CODE) {
+    if (PyObject_IsInstance(extra, (PyObject *)&CacheEntryWrapperType)) {
+      CacheEntry* e = ((CacheEntryWrapper*)extra)->cache_entry;
+      destroy_cache_entry(e);
+    } else if (PyDict_Check(extra)) {
+      PyObject* values = PyDict_Values(extra);
+      for (Py_ssize_t i = 0; i < PyList_GET_SIZE(values); i++) {
+        PyObject* value = PyList_GET_ITEM(values, i);
+        if (value != NULL && PyObject_IsInstance(value, (PyObject *)&CacheEntryWrapperType)) {
+          CacheEntry* e = ((CacheEntryWrapper*)value)->cache_entry;
+          destroy_cache_entry(e);
+        }
+      }
+    }
+  }
+  set_cache_entry_on_code((PyCodeObject*)code, NULL);
+}
+
+static void destroy_frame_state(PyObject* code) {
+  PyObject* frame_state = NULL;
+  _PyCode_GetExtra((PyObject*)code, dynamic_frame_state_extra_index, (void*)&frame_state);
+  Py_XDECREF(frame_state);
+  _PyCode_SetExtra((PyObject*)code, dynamic_frame_state_extra_index, NULL);
+}
+
 static PyObject* reset_code(PyObject* dummy, PyObject* code) {
   if (!PyCode_Check(code)) {
     DEBUG_TRACE0("arg error");
     PyErr_SetString(PyExc_TypeError, "expected a code object");
     return NULL;
   }
-
-  destroy_cache_entry(get_cache_entry((PyCodeObject*)code));
-  PyObject* frame_state = get_frame_state((PyCodeObject*)code);
-  Py_XDECREF(frame_state);
-  set_cache_entry((PyCodeObject*)code, NULL);
-  set_frame_state((PyCodeObject*)code, NULL);
+  destroy_cache_entry_wrapper(code);
+  destroy_frame_state(code);
```

Comment:
This PR is pretty big, consider separating these refactors out into their own prep PR?